### ISSUE TITLE
MacOS Packaging System

### ISSUE THREAD

**Issue by PeterPetrik** [qgis/QGIS-Enhancement-Proposals#143]
# QGIS Enhancement: MacOS Packaging System

**Date** 2019/03/22

**Author** Peter Petrik (@PeterPetrik)

**Contact** zilolv at gmail dot com

**maintainer** @PeterPetrik, @3nids

**Version** QGIS 3.8

# Summary

We have successfully created a [prototype](https://github.com/lutraconsulting/qgis-mac-packager) for community-driver packaging system for MacOS. The package is used by many QGIS users and the feedback is more than positive.The differences between the official (kyngchaos) package and the proposed package are the following:

- Supports Dark Mode (Mojave 10.14+)
- Single installer, all dependencies bundled in the package
- Signed installer and application
- Nightly builds on cloud MacOS server
- Allows installation of different versions side-by-side
- Transparent server setup, all scripts/docs on GitHub repository
- Embeds its own version of Python 3
- Embeds proj-datumgrids

The proposal states the steps needed to improve of the prototype to become official QGIS package tool.

# Repositories, Workflow and CI

We propose to have 2 separate repositories:

## Dependencies

First repository qgis/macos-deps contains Homebrew installation scripts to download and install all QGIS dependencies to /usr/local

- On each commit the CI (Continuous Integration, probably Travis CI) system will download homebrew configuration, install dependencies and package `/usr/local/Cellar` folder for further usage
- On request (e.g. git repository tagged) the CI (Continuous Integration) system will upload the dependency archive to a specific location for future usage of the packager

This separation is required to ensure the consistency of dependencies and the packaging code. The Homebrew deps are changing frequently (being source of instability) and it takes considerable time to install all of them on CI. 

## Packager

Packager qgis/macos-packager contains bash/python scripts to perform the  following:

- Fetch dependency archive from qgis/macos-deps
- Build QGIS LTR, PR or master from corresponding CI configuration
- Bundle all libraries, frameworks and shared data to QGIS's bundle
- Check and test the consistency of the package
- Create DMG installer
- Sign the package with QGIS Certificate
- If requested, upload the resulting package to OSGeo download area

# Tasks Breakdown

- Setup 2 different repositories. Transfer code from prototype to the repositories.
- Prepare scripts for deps repository to archive and test libraries from Homebrew for usage in other repository.
- Setup CI for qgis/macos-deps repository
- Document and refactor packager code to facilitate changes for future committers
- Setup CI for qgis/macos-packager repository
- Sign the bundle with QGIS Certificate (now signed with Lutra Consulting Certificate) on CI.
- Script to upload of the resulting package to download area and as Github release

**Comment by 3nids** [comment]
obviously 👍

**Comment by andreasneumann** [comment]
Hi, 

Peter: thanks for the summary of what needs to be done - and to reaching
out to Denis/OpenGIS to collaborate on the effort. 

About the package signing: do we really have a QGIS certificate or would
it rather be the OSGEO certificate? Or do you suggest to buy an
additional certificate for QGIS independent of the OSGEO one? 

I am in favour of supporting MacOS platform equally well like we support
the Windows and Linux platform (except for issues we can't control
ourselves (e.g. where qt has issues on MacOS)). 

QGIS' strength is, that it is available on multiple platforms. If we
give up this strength, I think it would be harmful to the project. There
are not too many other GIS available for the MacOS platform. 

Personally, I also think we should stop this attitude that MacOS users
should pay for the packages, while we don't ask our Windows or Linux
users to do the same. 

We have a budget of 9k EUR for packaging in 2019. In my opinion there
should be at least 3k EUR available for MacOS packaging out of this
"packaging budget". As far as I understand it, this should more or less
fill the gap that the MacOS crowd funding campaign didn't manage to
collect (not completely, but almost). 

It would be a shame not to build on the momentum of the LutraConsulting
effort and not to finish the process, started with the prototype. 

Let's put in on the Agenda of our upcoming PSC meeting and vote on it. 

Greetings, 

Andreas 

On 2019-03-22 15:29, Peter Petrik wrote:

> QGIS ENHANCEMENT: MACOS PACKAGING SYSTEM
> 
> DATE 2019/03/22 
> 
> AUTHOR Peter Petrik (@PeterPetrik [1]) 
> 
> CONTACT zilolv at gmail dot com 
> 
> MAINTAINER @PeterPetrik [1], @3nids [2] 
> 
> VERSION QGIS 3.8 
> 
> SUMMARY
> 
> We have successfully created a prototype [3] for community-driver packaging system for MacOS. The package is used by many QGIS users and the feedback is more than positive.The differences between the official (kyngchaos) package and the proposed package are the following: 
> 
> * Supports Dark Mode (Mojave 10.14+)
> * Single installer, all dependencies bundled in the package
> * Signed installer and application
> * Nightly builds on cloud MacOS server
> * Allows installation of different versions side-by-side
> * Transparent server setup, all scripts/docs on GitHub repository
> * Embeds its own version of Python 3
> * Embeds proj-datumgrids
> 
> The proposal states the steps needed to improve of the prototype to become official QGIS package tool. 
> 
> REPOSITORIES, WORKFLOW AND CI
> 
> We propose to have 2 separate repositories: 
> 
> DEPENDENCIES
> 
> First repository qgis/macos-deps contains Homebrew installation scripts to download and install all QGIS dependencies to /usr/local 
> 
> * On each commit the CI (Continuous Integration, probably Travis CI) system will download homebrew configuration, install dependencies and package /usr/local/Cellar folder for further usage
> * On request (e.g. git repository tagged) the CI (Continuous Integration) system will upload the dependency archive to a specific location for future usage of the packager
> 
> This separation is required to ensure the consistency of dependencies and the packaging code. The Homebrew deps are changing frequently (being source of instability) and it takes considerable time to install all of them on CI. 
> 
> PACKAGER
> 
> Packager qgis/macos-packager contains bash/python scripts to perform the following: 
> 
> * Fetch dependency archive from qgis/macos-deps
> * Build QGIS LTR, PR or master from corresponding CI configuration
> * Bundle all libraries, frameworks and shared data to QGIS's bundle
> * Check and test the consistency of the package
> * Create DMG installer
> * Sign the package with QGIS Certificate
> * If requested, upload the resulting package to OSGeo download area
> 
> TASKS BREAKDOWN
> 
> * Setup 2 different repositories. Transfer code from prototype to the repositories.
> * Prepare scripts for deps repository to archive and test libraries from Homebrew for usage in other repository.
> * Setup CI for qgis/macos-deps repository
> * Document and refactor packager code to facilitate changes for future committers
> * Setup CI for qgis/macos-packager repository
> * Sign the bundle with QGIS Certificate (now signed with Lutra Consulting Certificate) on CI.
> * Script to upload of the resulting package to download area and as Github release
> 
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub [4], or mute the thread [5].

 

Links:
------
[1] https://github.com/PeterPetrik
[2] https://github.com/3nids
[3] https://github.com/lutraconsulting/qgis-mac-packager
[4] https://github.com/qgis/QGIS-Enhancement-Proposals/issues/143
[5]

**Comment by nyalldawson** [comment]
@andreasneumann 

Thanks for the well thought out response. You've totally captured my thoughts here too.

**Comment by PeterPetrik** [comment]
I am easy with both QGIS or OSGEO certificate, whatever is easier to have/obtain.
Definitely we should aim for transparent and sustainable packaging process for all platforms.

**Comment by PeterPetrik** [comment]
One technical aspect: The packager itself does not rely (too much) from where the deps are taken from. The prototype is based on homebrew, but it is similarly possible to use conda (https://github.com/conda-forge/qgis-feedstock). Problably better maintained solution wins :)

**Comment by elpaso** [comment]
@PeterPetrik forgive my ignorance, if `conda install qgis` works well on Mac too, do you think it is still necessary to build an installer?

(the same argument would probably also apply to windows)

**Comment by jef-n** [comment]
The OSGeo certficate is available - and was originally bought to sign mac packages - although that apparently never did happen.

**Comment by luipir** [comment]
Hi Alesandro, conda is a data analysis oriented environment... A good
percentage (IMHO the majority) of of mac users are cartographers,
urbanologists, architects and... students (in US). None of these probably
know what is conda => a mac package is still necessary.

Luigi Pirelli

**************************************************************************************************
* LinkedIn: https://www.linkedin.com/in/luigipirelli
* Stackexchange: http://gis.stackexchange.com/users/19667/luigi-pirelli
* GitHub: https://github.com/luipir
* Mastering QGIS 2nd Edition:
*
https://www.packtpub.com/big-data-and-business-intelligence/mastering-qgis-second-edition
* Hire me: http://goo.gl/BYRQKg
**************************************************************************************************

**Comment by PeterPetrik** [comment]
Those are separate things. To install QGIS on Mac, you can do: 

- conda install qgis (basically creates you linux-like folder structure PREFIX/lib, PREFIX/bin with QGIS)
- brew install qgis (basically apt-like style. installs to /usr/local/bin, /use/local/)
- macports install qgis (detto)
- compile qgis yourself (not for users)
- install official packages (installs to /Application, but you need to install GDAL etc separately and not signed)
- install lutra's packages (install's to /Application. Style that Mac users are familiar with other tools. Similar to App Store)

So basically with conda/macports/brew you got unsigned packages that is created by downloading the dependencies one-by-one. The dependencies are not controlled well and updated frequently so you may end up with non-working installation. It is basically linux-style and most of the users do not have any familiarity with it (imagine you would like to explain windows users to understand /usr/local/lib and /usr/local/bin installation). It is not signed and tested properly so unlikely for big companies to use it.

**Comment by elpaso** [comment]
@PeterPetrik thanks for the detailed explanation!

**Comment by PeterPetrik** [comment]
@NathanW2 this one is accepted and implemented https://qgis.org/downloads/macos/
